provider/keystore/keystore.go

```diff
@@ -335,20 +402,35 @@ func (s *keystore) delete(ctx context.Context, keys []mh.Multihash) error {
 	if err != nil {
 		return err
 	}
+	seen := make(map[bit256.Key]struct{}, len(keys))
+	removedCount := 0
 	for _, h := range keys {
-		dsk := dsKey(keyspace.MhToBit256(h), s.prefixBits)
-		err := b.Delete(ctx, dsk)
+		k := keyspace.MhToBit256(h)
+		if _, ok := seen[k]; ok {
+			continue
+		}
+		seen[k] = struct{}{}
+		dsk := dsKey(k, s.prefixBits)
+		ok, err := s.ds.Has(ctx, dsk)
 		if err != nil {
 			return err
 		}
+		if ok {
+			if err := b.Delete(ctx, dsk); err != nil {
+				return err
+			}
+			removedCount++
+		}
 	}
+	s.size -= removedCount
 	return b.Commit(ctx)
 }
 
-// size returns the number of keys currently stored in the Keystore.
-func (s *keystore) size(ctx context.Context) (size int, err error) {
+// refreshSize iterates over all keys in the supplied datastore to count the
+// total number of stored keys.
+func refreshSize(ctx context.Context, d ds.Datastore) (size int, err error) {
```

Comment:
💭  not really a problem, but iiuc we are counting `sizeKey` here?  the result will be +1 which may confuse users who expect only 1 key or something :)

maybe
```go
if r.Key == sizeKey {
  continue
}
```
just to  be safe